File: week_1_sprint/two_pointers/python/two_sum_ii_input_array_is_sorted.py
# https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/description/

from typing import List

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        # Two Pointers
        # Time Complexity O(n)
        # Space Complexity O(1)
        i: int = 0
        j: int = len(numbers) - 1
        ans: List[int] = [-1, -1]

        while i < j:
            sum: int = numbers[i] + numbers[j]
            if (sum > target):
                j -= 1
            elif (sum < target):
                i += 1
            else:
                ans[0] = i + 1
                ans[1] = j + 1
                break

        return ans
            

File: leetcode/two-pointer-sliding-window/two-sum-ii-input-array-is-sorted.cpp
// https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/description/
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        // return my_two_pointers(numbers, target);
        
    }

    // My Soln
    // Two Pointers
    // Time Complexity O(n)
    // Space Complexity O(1)
    vector<int> my_two_pointers(const std::vector<int>& numbers, int target) {
        int n = numbers.size();
        int low = 0;
        int high = n - 1;

        while (low < high) {
            if (numbers[low] + numbers[high] == target) {
                return {low + 1, high + 1};
            }
            else if (numbers[low] + numbers[high] > target) {
                high--;
            }
            else {
                low++;
            }
        }
        return {-1, -1};
    }

};